src/tiledbsoma_ml/x_locator.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+from contextlib import contextmanager
+from typing import Dict, Generator, Tuple, Union
+
+import attrs
+from tiledbsoma import DataFrame, Experiment, SOMATileDBContext, SparseNDArray
+
+
+@attrs.define(frozen=True, kw_only=True)
+class XLocator:
+    """State required to open an ``X`` |SparseNDArray| (and associated ``obs`` |DataFrame|), within an |Experiment|.
+
+    Serializable across multiple processes.
+    """
+
+    uri: str
+    measurement_name: str
+    layer_name: str
+    tiledb_timestamp_ms: int
+    tiledb_config: Dict[str, Union[str, float]]
```

Comment:
stylistic nit picking:   prefer `Dict[str, str|float]`

Just requires an import at the top of file:
`from __future__ import annotations`